### PR TITLE
feat: [DEL-5792]: core changes to make container work

### PR DESCRIPTION
**Pull request by xingchi-jin**
## Describe your changes
This PR, together with https://github.com/harness/harness-core/pull/42635 will provide a runnable k8s ng task that can deploy nginx yaml

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43520)

### CODE REVIEW

File: 260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8SJob.java [hunk 1]
```diff
@@ -42,13 +46,15 @@ public K8SJob addVolume(final V1Volume volume, final V1VolumeMount volumeMount)
...
+  //container -> container.addVolumeMountsItem(volumeMount)
 
   public K8SJob addVolume(final V1Volume volume, final String mountPath) {
     final var volumeMount = K8SVolumeUtils.createVolumeMount(volume, mountPath);
     getSpec().getTemplate().getSpec().addVolumesItem(volume).getContainers().forEach(
         container -> container.addVolumeMountsItem(volumeMount));
-    getSpec().getTemplate().getSpec().getInitContainers().forEach(
-        container -> container.addVolumeMountsItem(volumeMount));
+    Optional.ofNullable(getSpec().getTemplate().getSpec().getInitContainers())
+        .orElse(Collections.<V1Container>emptyList())
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8SJob.java:56]: hit NPE when no init container is added

File: 260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8STaskRunner.java [hunk 2]
```diff
@@ -50,14 +54,20 @@ public class K8STaskRunner {
...
   private final BatchV1Api batchApi;
   private final CoreV1Api coreApi;
   private final DelegateConfiguration delegateConfiguration;
-  @Named("referenceFalseKryoSerializer") private final KryoSerializer kryoSerializer;
+  private final KryoSerializer kryoSerializer;
 
   @Inject
   public K8STaskRunner(final DelegateConfiguration delegateConfiguration, final ApiClient apiClient,
-      final KryoSerializer kryoSerializer) throws IOException {
+                       @Named("referenceFalseKryoSerializer") final KryoSerializer kryoSerializer) throws IOException {
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8STaskRunner.java:70]: guice doesn't honor the named reference if not adding to constructor

File: 260-delegate/src/main/java/io/harness/delegate/service/core/k8s/resources/k8s_task_template.yaml [hunk 3]
```diff
@@ -9,7 +9,7 @@ spec:
     spec:
       restartPolicy: Never
       containers:
-        - image: alpine
+        - image: harnessdev/delegate-runner:latest
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/delegate/service/core/k8s/resources/k8s_task_template.yaml:12]: temporarily use this image. until we figure out a way to dynamically embedded image path

File: 980-commons/src/main/java/io/harness/logging/BUILD.bazel [hunk 4]
```diff
@@ -10,6 +10,9 @@ java_library(
     name = "module",
     srcs = glob(["*.java"]),
     visibility = ["//visibility:public"],
+    resources = [
+        "//980-commons/src/main/resources:resource",
```
Comment (xingchi-jin) [inline 980-commons/src/main/java/io/harness/logging/BUILD.bazel:14]: this is a fix regardless of bijou project.

**Review by mdmilic — APPROVED** [review]
> Reviewed 31 of 32 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43520)* status: Some of the checks are too old. Waiting on team(s): del-team. 

___
*[`260-delegate/dw-config-delegate.yml` line 3 at r2](https://reviewable.io/reviews/harness/harness-core/43520#-NNkzADKABKwUQkiNnUf:-NNkzADKABKwUQkiNnUg:b-gnuthk) ([raw file](https://github.com/harness/harness-core/blob/f95f5373f0d4e594cf521e49573047e5730beb95/260-delegate/dw-config-delegate.yml#L3)):*
> ```YAML
> # Dropwizard config
> server:
>   allowedMethods: ["GET", "POST"]
> ```

Note, this only applies to running delegate from intellij. We would need to update the config that goes into the delegate container. Currently immutable delegate config is reused, but for immutable we don't want to allow POST, so we should create separate config. You can do it in separate PR to not block this one. 
This is the file used in the container
https://github.com/harness/harness-core/blob/develop/dockerization/delegate/immutable-scripts/config.yml
___
*[`260-delegate/src/main/java/io/harness/delegate/app/modules/platform/BUILD.bazel` line 24 at r2](https://reviewable.io/reviews/harness/harness-core/43520#-NNl-5aS79w-59OMeAiB:-NNl-5aT7gYPAfs10_im:b-e1nq91) ([raw file](https://github.com/harness/harness-core/blob/f95f5373f0d4e594cf521e49573047e5730beb95/260-delegate/src/main/java/io/harness/delegate/app/modules/platform/BUILD.bazel#L24)):*
> ```Starlark
>         "//940-secret-manager-client:module",
>         "//950-delegate-tasks-beans:module",
>         "//930-delegate-tasks:module",
> ```

why this :cry: 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/CoreDelegateService.java` line 44 at r2](https://reviewable.io/reviews/harness/harness-core/43520#-NNl2AiB96Yg3AXVPlK0:-NNl2AiC47C-Je-hx_Hm:b8smo6h) ([raw file](https://github.com/harness/harness-core/blob/f95f5373f0d4e594cf521e49573047e5730beb95/260-delegate/src/main/java/io/harness/delegate/service/core/CoreDelegateService.java#L44)):*
> ```Java
>   protected void executeTask(final @NonNull DelegateTaskPackage taskPackage) {
>     try {
>       // FIXME: remove task type restriction
> ```

FYI the way to do this is actually to only announce tasks your delegate supports, manager then makes sure it only sends the right task types. This is done in 

https://github.com/harness/harness-core/blob/develop/260-delegate/src/main/java/io/harness/delegate/service/core/CoreDelegateService.java#L60

We can leave this for now though and remove later. 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8STaskRunner.java` line 179 at r2](https://reviewable.io/reviews/harness/harness-core/43520#-NNl4ZUp4DOgzuOCk6Zi:-NNl4ZUp4DOgzuOCk6Zj:b-bave0t) ([raw file](https://github.com/harness/harness-core/blob/f95f5373f0d4e594cf521e49573047e5730beb95/260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8STaskRunner.java#L179)):*
> ```Java
> 
>   private V1ConfigMap createTaskExecutorConfig(final String taskId) throws IOException, ApiException {
>     var configBuilder =
> ```

Delegate should not be aware of executor. Executor exists only for harness tasks as a helper lib, not for 3rd party tasks. Any config that needs to be passed to the executor needs to go through task agent config. We can change that later. 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8STaskRunner.java` line 186 at r2](https://reviewable.io/reviews/harness/harness-core/43520#-NNl3_ot9xseC38p_rmx:-NNl3_ot9xseC38p_rmy:bb2kp9g) ([raw file](https://github.com/harness/harness-core/blob/f95f5373f0d4e594cf521e49573047e5730beb95/260-delegate/src/main/java/io/harness/delegate/service/core/k8s/K8STaskRunner.java#L186)):*
> ```Java
>             .delegateName(DELEGATE_NAME);
> 
>     // FIXME: Use service based fqdn for kubernetes and helm delegates. Use ip for other delegates
> ```

Executor shouldn't need this at all. The service endpoint is available automatically by K8S. For other envs like VMs we can simulate that (e.g. add env vars manually), but that's out of scope now. 



_Code quote:_
```Java
    // FIXME: Use service based fqdn for kubernetes and helm delegates. Use ip for other delegates
    configBuilder.delegateHost(Localhost.getLocalHostAddress()).delegatePort(3460);
```

**Pull request closed without merging**